galley/pkg/config/analysis/analyzers/auth/jwt.go

```diff
@@ -0,0 +1,170 @@
+		j.analyzeServiceTarget(r, c, nsm)
+		return true
+	})
+}
+
+// buildNamespaceServiceMap returns a map where the index is a namespace and the boolean
+func (j *JwtAnalyzer) buildNamespaceServiceMap(ctx analysis.Context) map[string]*v1.ServiceSpec {
+	// Keep track of each fqdn -> service definition
+	fqdnServices := map[string]*v1.ServiceSpec{}
+
+	ctx.ForEach(collections.K8SCoreV1Services.Name(), func(r *resource.Instance) bool {
+		svcNs := r.Metadata.FullName.Namespace
+		svcName := r.Metadata.FullName.Name
+
+		svc := r.Message.(*v1.ServiceSpec)
+		fqdn := util.ConvertHostToFQDN(svcNs, string(svcName))
+		fqdnServices[fqdn] = svc
+
+		return true
+	})
+
+	return fqdnServices
+}
+
+func (j *JwtAnalyzer) analyzeServiceTarget(r *resource.Instance, ctx analysis.Context, nsm map[string]*v1.ServiceSpec) {
+	policy := r.Message.(*v1alpha1.Policy)
+	ns := r.Metadata.FullName.Namespace
+
+	for _, origin := range policy.Origins {
+		if origin.GetJwt() != nil {
+			for _, target := range policy.GetTargets() {
+
+				fqdn := util.ConvertHostToFQDN(ns, target.GetName())
+				if svc, has := nsm[fqdn]; has {
+					if len(target.GetPorts()) == 0 {
+						checkServicePorts(r, ctx, svc)
+						continue
+					}
+
+					// check ports defined in the authentication policy for the service
+					for _, port := range target.GetPorts() {
+						if port.GetName() == "" {
+							checkPort(r, ctx, port.GetNumber(), svc)
+						}
+
+					}
+				} // else no target service was found, but not an error
+			}
+		}
+	}
+
+}
+
+func checkPort(r *resource.Instance, ctx analysis.Context, portNum uint32, svc *v1.ServiceSpec) {
+	var svcPort *v1.ServicePort
+	for _, port := range svc.Ports {
+		if strings.ToUpper(string(port.Protocol)) != "TCP" && port.Protocol != "" {
+			continue
+		}
+		if portNum == uint32(port.Port) {
+			svcPort = &port
+			break
+		}
+	}
+	if svcPort != nil {
+		portName := strings.ToLower(svcPort.Name)
+		if !portHasSupportedPrefix(portName) {
+			ctx.Report(collections.IstioAuthenticationV1Alpha1Policies.Name(),
+				msg.NewJwtFailureDueToInvalidServicePortPrefix(
+					r,
+					int(portNum),
+					portName,
+					string(svcPort.Protocol),
+					svcPort.TargetPort.String(),
+				))
+		}
+
+	}
+}
+
+func checkServicePorts(r *resource.Instance, ctx analysis.Context, svc *v1.ServiceSpec) {
+	for _, port := range svc.Ports {
+		portName := strings.ToLower(port.Name)
+		if (strings.ToUpper(string(port.Protocol)) == "TCP" || port.Protocol == "") &&
+			portHasSupportedPrefix(portName) {
+			continue
+		} else {
+			ctx.Report(collections.IstioAuthenticationV1Alpha1Policies.Name(),
+				msg.NewJwtFailureDueToInvalidServicePortPrefix(
+					r,
+					int(port.Port),
+					portName,
+					string(port.Protocol),
+					port.TargetPort.String(),
+				))
+
+		}
+	}
+}
+
+func portHasSupportedPrefix(portName string) bool {
+	return strings.HasPrefix(portName, portPrefixHttp) ||
+		strings.HasPrefix(portName, portPrefixHttp2) ||
```

Comment:
It is unclear to me if `grpc` and `grpc-web` are also supported. I would like someone to clarify this for me. I can try and do some testing locally with a grpc echo client/server.

Reply:
@lei-tang can you help here?

Reply:
@diemtvu I think this PR may be related to your JWT authn policy work https://docs.google.com/document/d/1f1ZVCUa7ktuoFKKFATCXfu9qFFFXX-7UMAJQtmMPuS4.